### PR TITLE
Legion: Add hooks & types for accessing new data model

### PR DESCRIPTION
**Pull request by msnyder-msft**
### Summary of changes 🔍 
Adding new types for defining the DB and ViewModels for the app level data for the following items:
- `Entity`
- `Type`
- `Relationship`
- `RelationshipType`
- `Property`

Each type has a base interface (ex: `IEntityBase`) which contains the common attributes of the item. Then there is a DB version with the foreign key references and the View version which has the child items resolved into full objects.

image

We also introduced a new `WizardDataContext` that hosts these entities in their denormalized structure with foreign key references. Then we use hooks to access that context and hydrate each type of entity so that components can use the data directly without needing to link data across the different entities.
- `useEntities`
- `useTypes`
- `useRelationships`
- `useProperties`

Each hook exposes methods for doing updates on that kind of entity and will push it back into the global state.

NOTE: the callback implementations are not deep operations. We will have to decide which callbacks will propogate changes to the child items. Ex: if you update a `property` on a `type` of an `entity` do we reflect that everywhere? What's the right model for doing that and the cross dependency between hooks? For now, this PR will have a naive implementation and will give us the types and functions to call. The actual implementations can get revved in future PRs.

### Testing 🧪
This has not yet been consumed anywhere.

### CODE REVIEW

File: src/Apps/Legion/Models/AppTypes.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
+/**
+ * This file is for containing the top level types for use across the entire app. It should not contain any types specific to particular components or segments of the app.
```
Comment (iremgokce) [inline src/Apps/Legion/Models/AppTypes.ts:2]: is there a reason why this is called "AppTypes" but not just "Types"?

File: src/Apps/Legion/Services/AppTypeUtilities.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        friendlyName: viewModel.friendlyName,
+        id: viewModel.id,
+        sourceConnectionString: viewModel.sourceConnectionString,
+        sourceEntityId: viewModel.sourceEntityId,
+        values: viewModel.values, // TODO: do we need to serialize each property value into a string?
+        typeId: viewModel.type.id
+    };
+}
+
+export function convertEntityToView(
```
Comment (iremgokce) [inline src/Apps/Legion/Services/AppTypeUtilities.ts:26]:
```suggestion
export function convertDbEntityToView(
```

File: src/Apps/Legion/Services/AppTypeUtilities.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    IDbType,
+    IViewEntity,
+    IViewProperty,
+    IViewRelationship,
+    IViewRelationshipType,
+    IViewType
+} from '../Models';
+
+// #region Entities
+export function convertEntityToDb(viewModel: IViewEntity): IDbEntity {
```
Comment (iremgokce) [inline src/Apps/Legion/Services/AppTypeUtilities.ts:15]:
```suggestion
export function convertViewEntityToDb(viewModel: IViewEntity): IDbEntity {
```

File: src/Apps/Legion/Services/AppTypeUtilities.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+export function convertEntityToView(
+    dbModel: IDbEntity,
+    state: {
+        types: IDbType[];
+        properties: IDbProperty[];
+    }
+): IViewEntity {
+    const type = state.types.find((x) => x.id === dbModel.typeId);
+    return {
+        friendlyName: dbModel.friendlyName,
```
Comment (iremgokce) [inline src/Apps/Legion/Services/AppTypeUtilities.ts:35]: same comment for the following methods to use spread method to copy the same fields to the new object except some field

File: src/Apps/Legion/Services/AppTypeUtilities.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    IDbRelationshipType,
+    IDbType,
+    IViewEntity,
+    IViewProperty,
+    IViewRelationship,
+    IViewRelationshipType,
+    IViewType
+} from '../Models';
+
+// #region Entities
```
Comment (iremgokce) [inline src/Apps/Legion/Services/AppTypeUtilities.ts:14]: Same question here, why not just "TypeUtilities" as the name of the file?

**Comment by iremgokce** [comment]
Ahh it is just got merged, please take a look at @msnyder-msft if any of those make sense moving forward.

File: src/Apps/Legion/Models/AppTypes.ts (hunk 1 above)
Comment (msnyder-msft) [reply src/Apps/Legion/Models/AppTypes.ts:2]: Yeah. I wanted a file just for the data types we use for the wizard. Maybe I should have called it Wizard.types.ts.

File: src/Apps/Legion/Services/AppTypeUtilities.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+export function convertEntityToDb(viewModel: IViewEntity): IDbEntity {
+    return {
+        friendlyName: viewModel.friendlyName,
```
[the remark this answers is not in the document]
Comment (msnyder-msft) [reply src/Apps/Legion/Services/AppTypeUtilities.ts:17]: I thought about it, but decided not to. Spread will also include other data in the type even if it's not on the type definition so I wanted to have a clean conversion that doesn't bring anything extra in.

**Pull request merged**